api/src/main/java/edu/cornell/mannlib/vitro/webapp/rdfservice/impl/sparql/RDFServiceSparql.java

```diff
@@ -384,32 +383,45 @@ public boolean sparqlAskQuery(String queryStr) throws RDFServiceException {
 		}
 	}
 
-	/**
-	 * Get a list of all the graph URIs in the RDF store.
-	 */
-	@Override
-	public List<String> getGraphURIs() throws RDFServiceException {
-		if (graphURIs == null || rebuildGraphURICache) {
-			graphURIs = getGraphURIsFromSparqlQuery();
-			rebuildGraphURICache = false;
-		}
-		return graphURIs;
-	}
-
-	private List<String> getGraphURIsFromSparqlQuery() throws RDFServiceException {
-		String fastJenaQuery = "SELECT DISTINCT ?g WHERE { GRAPH ?g {} } ORDER BY ?g";
-		String standardQuery = "SELECT DISTINCT ?g WHERE { GRAPH ?g { ?s ?p ?o } }";
-		List<String> graphURIs = new ArrayList<String>();
-		try {
-			graphURIs = getGraphURIsFromSparqlQuery(fastJenaQuery);
-		} catch (Exception e) {
-			log.debug("Unable to use non-standard ARQ query for graph list", e);
-		}
-		if (graphURIs.isEmpty()) {
-			graphURIs = getGraphURIsFromSparqlQuery(standardQuery);
-			Collections.sort(graphURIs);
-		}
-		return graphURIs;
+	protected void rebuildGraphUris() {
+        Thread thread = new VitroBackgroundThread(new Runnable() {
+            public void run() {
+                synchronized (RDFServiceJena.class) {
+                    if (rebuildGraphURICache) {
+                        try {
+                            isRebuildGraphURICacheRunning = true;
+                            Set<String> newGraphUris = new HashSet<>();
+                            try {
+                                String fastJenaQuery = "SELECT DISTINCT ?g WHERE { GRAPH ?g {} } ORDER BY ?g";
+                                newGraphUris.addAll(getGraphURIsFromSparqlQuery(fastJenaQuery));
+                            } catch (Exception e) {
+                                log.debug("Unable to use non-standard ARQ query for graph list", e);
+                            }
+                            if (graphURIs.isEmpty()) {
+                                String standardQuery = "SELECT DISTINCT ?g WHERE { GRAPH ?g { ?s ?p ?o } }";
+                                newGraphUris.addAll(getGraphURIsFromSparqlQuery(standardQuery));
+                            }
+                            Set<String> oldGraphUris = new HashSet<String>(graphURIs);
+                            if (newGraphUris.equals(oldGraphUris)) {
+                                return;
+                            }
+                            Set<String> removedGraphUris = new HashSet<String>(oldGraphUris);
+                            removedGraphUris.removeAll(newGraphUris);
+                            graphURIs.removeAll(removedGraphUris);
+                            Set<String> addedGraphUris = new HashSet<String>(newGraphUris);
+                            addedGraphUris.removeAll(oldGraphUris);
+                            graphURIs.addAll(addedGraphUris);
```

Comment:
This snippet of code is the same as in RDFServiceJena. I am wondering whether we can move that in a method in RDFServiceImpl abstract class.

Reply:
Done